### PR TITLE
Fix TestForEmptyStringsUsingStringLength.Fixer when empty literal is used

### LINKED ISSUE #3668: Refactor for string IsNullOrEmpty
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/1034835/refactor-for-string-isnullorempty.html)._

---
Given the following code:

```
        public static bool Simulate => s_pCartId == "";
```

The IDE warns that string. Length or string. IsNullOrEmpty should be used instead of s_pCartId == "". Unfortunately the lightbulb pulldown does not provide any selections to perform this operation automatically which results in a lot of manual typing to resolve.

---
#### Visual Studio Feedback System on 5/18/2020, 01:16 AM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

### PR DESCRIPTION
**Pull request by Youssef1313**
The `ContainsSystemStringEmpty` was returning true for `string.Empty`, but not for `""`. The change should return true on both cases.

Fixes #3668

### CODE REVIEW

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs [hunk 1]
```diff
@@ -37,5 +37,10 @@ protected override SyntaxNode GetRightOperand(SyntaxNode binaryExpressionSyntax)
         {
             return ((BinaryExpressionSyntax)binaryExpressionSyntax).Right;
         }
+
+        protected override bool ContainsEmptyStringLiteral(SyntaxNode node)
+        {
+            return node is LiteralExpressionSyntax literalExpressionSyntax && literalExpressionSyntax.Token.ValueText.Length == 0;
+        }
```
Comment (Evangelink) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs:44]: [cosmetic] Might be worth using the arrowed body here instead.

```suggestion
        => node is LiteralExpressionSyntax literalExpressionSyntax &&
        	literalExpressionSyntax.Token.ValueText.Length == 0;
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs [hunk 2]
```diff
@@ -14,6 +14,42 @@ namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
     public class TestForEmptyStringsUsingStringLengthFixerTests
     {
         private const int c_StringLengthCodeActionIndex = 1;
+
+        [Fact]
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs:18]:
```suggestion
        [Fact, WorkItem(3686, "https://github.com/dotnet/roslyn-analyzers/pull/3686")]
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs [hunk 3]
```diff
@@ -14,6 +14,42 @@ namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
...
+        public async Task CA1820_FixTestEmptyStringsUsingIsNullOrEmpty_WhenStringIsLiteral()
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs:19]: Could you add a test with the other fixer too?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs [hunk 4]
```diff
@@ -37,5 +37,10 @@ protected override SyntaxNode GetRightOperand(SyntaxNode binaryExpressionSyntax)
         {
             return ((BinaryExpressionSyntax)binaryExpressionSyntax).Right;
         }
+
+        protected override bool ContainsEmptyStringLiteral(SyntaxNode node)
+        {
+            return node is LiteralExpressionSyntax literalExpressionSyntax && literalExpressionSyntax.Token.ValueText.Length == 0;
```
Comment (Evangelink) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs:43]: @mavasani Is it possible to have null for the `ValueText`?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs (hunk 3 above)
Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs:19]: Also, could you add 2 tests with the empty string on the left side? I am fine with adding the cases on the existing test(s).

Comment (Youssef1313) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLengthTests.Fixer.cs:19]: I'll add them soon.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs [hunk 5]
```diff
@@ -37,5 +37,9 @@ protected override SyntaxNode GetRightOperand(SyntaxNode binaryExpressionSyntax)
         {
             return ((BinaryExpressionSyntax)binaryExpressionSyntax).Right;
         }
+
+        protected override bool ContainsEmptyStringLiteral(SyntaxNode node)
```
Comment (mavasani) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs:41]: I think this can be implemented in the base type using [SemanticModel.GetConstantValue](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Compilation/SemanticModel.cs,744) API

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLength.Fixer.cs [hunk 6]
```diff
@@ -140,6 +140,7 @@ private async Task<Document> ConvertToStringLengthComparison(CodeFixContext cont
         protected abstract bool IsNotEqualsOperator(SyntaxNode node);
         protected abstract SyntaxNode GetLeftOperand(SyntaxNode binaryExpressionSyntax);
         protected abstract SyntaxNode GetRightOperand(SyntaxNode binaryExpressionSyntax);
+        protected abstract bool ContainsEmptyStringLiteral(SyntaxNode node);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLength.Fixer.cs:143]:
```suggestion
        private bool ContainsEmptyStringLiteral(SyntaxNode node, SemanticModel model, CancellationToken cancellationToken)
            => model.GetConstantValue(node, cancellationToken) is optionalValue && 
               optionalValue.HasValue &&
               optionalValue.Value is string value &&
               value.Length == 0;
```

**Comment by Youssef1313** [comment]
@mavasani, Sure. Will complete it very soon.
Thanks for the review.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLength.Fixer.cs [hunk 7]
```diff
@@ -55,18 +56,18 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
 
-        private FixResolution? TryGetFixResolution(SyntaxNode binaryExpressionSyntax, SemanticModel model)
+        private FixResolution? TryGetFixResolution(SyntaxNode binaryExpressionSyntax, SemanticModel model, CancellationToken cancellationToken)
         {
             bool isEqualsOperator = IsEqualsOperator(binaryExpressionSyntax);
             SyntaxNode leftOperand = GetLeftOperand(binaryExpressionSyntax);
             SyntaxNode rightOperand = GetRightOperand(binaryExpressionSyntax);
 
-            if (ContainsSystemStringEmpty(leftOperand, model))
+            if (ContainsSystemStringEmpty(leftOperand, model) || ContainsEmptyStringLiteral(leftOperand, model, cancellationToken))
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/TestForEmptyStringsUsingStringLength.Fixer.cs:65]: @mavasani, I think the `ContainsSystemStringEmpty` is now unneeded.
The implementation you referred to for the `ContainsEmptyStringLiteral` checks for a constant value of empty string whatever its source (literal, System.String.Empty, or another constant).

We may remove the `ContainsSystemStringEmpty`, then rename `ContainsEmptyStringLiteral` to be more descriptive to `ContainsEmptyString`, right?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs [hunk 8]
```diff
@@ -38,4 +38,4 @@ protected override SyntaxNode GetRightOperand(SyntaxNode binaryExpressionSyntax)
             return ((BinaryExpressionSyntax)binaryExpressionSyntax).Right;
         }
     }
-}
\ No newline at end of file
+}
```
Comment (mavasani) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpTestForEmptyStringsUsingStringLength.Fixer.cs:41]: Revert?
